### ISSUE TITLE
Allow for a way to mutate URLRequests before they're sent

### ISSUE THREAD

**Issue by SilverTab** [borchero/Squid#3]
Some context: I'm working with an API that requires a header field that is a hash computed with several elements of the http request (the POST body, the path, etc.).

I was able to do it by forking and making a couple of internal structs public (`HttpRequest` and the Network Scheduler), but it feels rather hack-ish, not quite sure what a good implementation would be, will try to think of a nicer implementation, but in the meantime, thought I'd let you know that this is a feature that some people might need. Loving the library so far BTW, nice work!

**Comment by borchero** [comment]
I’ll think about how to do that, stay tuned ;)

**Comment by SilverTab** [comment]
Thanks, I was able to do it without having to fork actually, but it's still a rather inelegant solution. The main difficulty stemmed from the fact that I can't get the raw values for HttpQuery and HttpRoute in order to compute the header, because their raw values are marked as internal. I was able to get around that limitation by doing this:

```
protocol AuthenticatedRequest: JsonRequest {
    var rawRoutes: [String] { get }
    var rawQuery: [String: String] { get }
}
extension AuthenticatedRequest {
    public var routes: HttpRoute {
        return HttpRoute(rawRoutes)
    }
    public var query: HttpQuery {
        return HttpQuery(rawQuery)
    }
```

But... yeah, feels like the wrong way of doing it haha!

**Comment by borchero** [comment]
I'm thinking about adding a function with the following signature to the underlying `Request` protocol:

```swift
func prepare(_ request: URLRequest) -> URLRequest
```

This function is passed the URL request that is going to be sent and, by default, simply returns the request being passed. What do you think about that?

**Comment by SilverTab** [comment]
Much better! 😀

**Comment by borchero** [comment]
It's published as version `1.1.2` ;)
